README

Packet-o-matic : the modular network packet processor.
Author : Guy Martin <gmsoft@tuxicoman.be>
Homepage : http://www.packet-o-matic.org
Additional documentation : http://wiki.packet-o-matic.org


What it is :
------------
Packet-o-matic is a modular network sniffer and packet processor. The modules consists of input, match, conntrack, helper and target.
It is designed to be as protocol independant as possible. Each protocol has it's own module for match, conntrack and helper (if applicable).


Building from svn :
-------------------
You need at least autoconf-2.61. First run "autoreconf -f -i" to create ./configure and then follow the instructions as if you fetched from a tarball.


Building from tarball :
-----------------------
Run these commands in order :
./configure
make
make install


Running it :
------------
If you make install, you are fine. The program will be able to find the libraries using automatic resolving or hardcoded library path.
If you did not run make install, you have to `export LD_LIBRARY_PATH=path/to/libdir`. By default, libtool creates the libs into the src/.libs directory.

Some modules such as input_pcap requires root privileges in order to perform their operation.


Dependencies :
--------------
 - libpcap : required to build input_pcap, target_pcap, target_inject and target_tcpkill
 - libxml2 : mandatory to parse the configuration
 - xmlrpc-c : required to build XML-RPC interface and web interface
 - MySQL : required to build datastore_mysql
 - PostgreSQL : required to build datastore_postgres
 - SQLite3 : required to build datastore_sqlite


How it works :
--------------
Packets are read from the input.
We check against each match of each rule if we should make that packet processed by the targets of the rule.
At the same time we check if a helper is needed and for example put all the ipv4 fragments into a single packet.
We check if the packet has some conntrack information attached to it.
If the packet has conntrack info, it is processed by the coresponding targets which saved the conntrack info.
It the match result is positive, we process the packet with the targets defined for this rule.


Getting started :
-----------------
First, start packet-o-matic by running `packet-o-matic` in the console.
Once started, it will listen on port 4655. Simply telnet to that port and start issuing commands.
You have to perform the following tasks to get a running configuration :
 - choose an input type and configure it
 - add the rules you need
 - add the targets you need
 - save the config

The following example start an input pcap on eth0 (the default) and dump all the images that goes over HTTP on port 80 :

// we use input pcap
pom> input type set pcap

// we want to match everything destinated to the port 80
pom> rule add tcp.dport == 80
Added rule with id 0

// the rule needs to be enabled
pom> rule enable 0

// target_http will dump part of the http traffic into files
pom> target add 0 http
Added target with id 0 to rule 0

// enable dumping the images
pom> target parameter set 0 0 dump_img yes

// start the target
pom> target start 0 0

// finaly start the input to capture packets and start dumping
pom> input start

// to save our changes we simply write the configuration
pom> config write
Configuration written in pom.xml.conf


The inputs :
------------
The input modules read packet from a certain source and give the packets to the rule processor.
A input has one or multiple mode. Each mode has its own set of parameters.

- pcap : allow you to read from one or all network device or from a pcap file saved by tcpdump or ethereal for example
- docsis : allow you to read fron a docsis stream using a dvb-c card and the linux-dvb api


The matchs :
------------
The match modules are used to specify the network traffic that you want to match and pass to the targets.

IMPORTANT : You'll have to specify the full header chain to match a packet. For example to match tcp traffic (on IPv4), you have to match in order ethernet, ipv4, tcp.

- docsis : match the docsis protocol used by cable modems
- ethernet : match the ethernet protocol
- ipv4 : match the IPv4 protocol commonly known as ip
- ipv6 : match the IPv6 protocol
- icmp : match the ICMP protocol
- icmpv6 : match the ICMPv6 protocol
- linux_cooked : match linux cooked protocol used by pcap
- rtp : match the RTP protocol used to carry real time voice and video
- tcp : match the TCP protocol
- udp : match the UDP protocol
- undefined : used internaly. do not use in your config !
- vlan : match 802.1Q aka vlan frames


The targets :
-------------
The target modules process the packet and most of the time, saves it or resend it.

- display : display usefull information about the packets being matched
- dump_payload : saves the payload of the last matching protocol into a file
- inject : reinject the matched packets on an specific interface
- irc : dump IRC connection into separate files with irssi-like log format
- pcap : save the packets in a pcap file
- pop : save emails and logins from POP connections into maildir
- rtp : save RTP payload into .au files. Supported payload types are G.711U, G711A, G.721 and G.722
- tap : open a tap interface and send the packets trough it
- tcpkill : works only with tcp packets. Send a TCP RST packet to any matched tcp packet
- tftp : dump files downloaded/uploaded via TFTP
- msn : dump conversations and avatar out of msn conversations
- null : does nothing, used for debugging or as a template


The conntracks :
----------------
The conntrack modules make the application able to recognize different connections among all the packets that it process.
It makes it easy to follow a connection and do something specific with that connection or save some information related to that connection.
The conntrack modules are automatically loaded if found, when the corresponding match is loaded.


The helpers :
-------------
The helper modules are used to recompose part of network packets which were splitted or modified by the network.
Helpers are processed before any matching is done with the rules.
The helper module currently implemented are ipv4, tcp and rtp. TODO : ipv6.
 
- ipv4 : perform ip fragment reassembly
- tcp : perform tcp reordering
- rtp : perform rtp reordering


Bugs :
------
Please report bugs to <gmsoft@tuxicoman.be> with version info and os + architercture.
If it's a crash, please provide a backtrace or a corefile.


License :
--------
packet-o-matic : modular network traffic processor
Copyright (C) 2007 Guy Martin <gmsoft@tuxicoman.be>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA



ChangeLog
* 2011/08/22 Guy Martin <gmsoft@tuxicoman.be>
Add filter_docsis3 parameter to input_docsis to drop docsis 3 packets when sniffing with only one card.
Don't cleanup automatically cleaned up libxml2 stuff.
Fix unserialization of ptype_timestamp.

* 2011/05/26 Guy Martin <gmsoft@tuxicoman.be>
Use bytea instead of varchara for strings handling in datastore_postgres and get rid of the UTF8 checks.
Slight optimization in http log handling.

* 2011/02/24 Guy Martin <gmsoft@tuxicoman.be>
Add expansion %{time.<strftime_format>} for filenames.
Fix memory leak when the input thread stops because of EOF thanks to Adrien Kunysz <adrien@kunysz.be>.
Fix possible buffer overflow in layer_field_parse().

* 2011/01/12 Guy Martin <gmsoft@tuxicoman.be>
Added header checksum validation in input_docsis.
Reorganized perf counters in input_docsis.
Update debug messages in input_docsis.

* 2011/01/02 Guy Martin <gmsoft@tuxicoman.be>
Added partial support for MSNP21.
Added some more error checking in target_msn.
Fix possible segfault in target_msn.

* 2010/11/25 Guy Martin <gmsoft@tuxicoman.be>
Remove param eurodocsis from input_docsis and deduce the type from the card frontend type.
Print all frequencies present in MDD when input_docsis is not in docsis3 mode.

* 2010/10/31 Guy Martin <gmsoft@tuxicoman.be>
Add async_commit parameter to datastore_postgres.

* 2010/10/28 Guy Martin <gmsoft@tuxicoman.be>
Fix database logging when a request in the same connection has alreay been logged.
Fix perf warning when pom is invoked incorrectly.
Fix small memory leak and behavior when listing modules if path is not found.
Add parsed_transactions to target_http.
Fix buffer handling in target_http on connection state change.
Handle all HTTP method and not only POST and GET in target_http.
Fix parse_errors counter in target_http.

* 2010/10/20 Guy Martin <gmsoft@tuxicoman.be>
Fix parsing log_format for target_http.
Fix logging when content-type is provided in the request.
Validate strings for UTF-8 in datstore_postgres.

* 2010/09/18 Guy Martin <gmsoft@tuxicoman.be>
Add some video mime types in mime_types.db.
Fix possible crash with extended headers in input_docsis.
Recoded target_irc.
Fix return code of xmlrpccmd_set_input_type() in case of error.
Fix compilation on OSX.
Return false for invalid operations in ptype_compare_val().
Update autoreconf command in README.

* 2010/06/21 Guy Martin <gmsoft@tuxicoman.be>
Fix crash in target_pop when user command ends with \n and not \r\n.
Fix crash in target_msn when no conversation is known while processing a INVITE.
Fix month handling in datastore_mysql.

* 2010/02/27 Guy Martin <gmsoft@tuxicoman.be>
Save uid of targets/rules/datastores in the config and restore them.
Added API to handle uids easily.
Fix browsing oid of common perf targets when a rule has no target.
Added extra perfmon items for target_tftp.
Fix usage of write() in target_tftp.

* 2010/02/23 Guy Martin <gmsoft@tuxicoman.be>
Tune to all frequencies at once in DOCSIS MDD messages.
Fix uncX counter for input_docsis.
Fix snmp getnext processing for target perf when some target don't have extra items.
Fix usage of write() in target_tap.
Reset extra perfmon counters for input and targets as well as known ones when needed.
Allow receiving presence for ourself with NLN messages in target_msn.
Added extra perfmon items for target_pop.

* 2010/02/21 Guy Martin <gmsoft@tuxicoman.be>
Prevent crash when specifying an invalid target uid with snmp get.
Sanitize written file names in target_msn.
Added performance items to target_msn.
Remove MSN buddies from the hash list once they don't belong to any session anymore.
Updated copyright and added small note for help for mgmt cli.

* 2010/02/17 Guy Martin <gmsoft@tuxicoman.be>
Check for timeout when waiting for packets from all adapters in input_docsis.
Fix SNMP target perf objects when more than one perf object exists.
Use write() correctly in target_http.
Added performance items in target_http.

* 2010/02/08 Guy Martin <gmsoft@tuxicoman.be>
Fix ioctl call to FE_READ_BER in input_docsis.
Update SNMP MIB for perfs.
Added OIDs for target perf objects.
Fix xmlrpccmd_get_target() broken in last commit.
Fix timestamp handling on big endian platform for datastore_postgres.
Added extra perf objects to target_dump_payload.
Improve locking of perf objects.	
Remove now useless performance stats on target_inject.

* 2010/01/23 Guy Martin <gmsoft@tuxicoman.be>
Fix signature of xmlrpc function rules.remove.
Fix crash in target snmp oids when no rules are configured.
Simplify locking handling in xmlrpccmd_get_target().
Added XML-RPC functions for datastores.

* 2010/01/09 Guy Martin <gmsoft@tuxicoman.be>
No need to check for FE_GET_EVENT in input_docsis.
Do not check for SYNC messages on non-primary capable DOCSIS channels.
Added perf_item_val_uptime_restart().
Added locks to struct perf_instance and struct perf_item.
Fix compilation warnings on 64bit platform for perf.c.

* 2009/12/23 Guy Martin <gmsoft@tuxicoman.be>
Added dropped packets perf item to input_pcap.
Added serial oids for input, rules and targets.
Update all the serials when input starts the target on load.
Don't unlock the rbuf mutex if the input is running when updating input param via SNMP.
Use bigger buffers for displaying perf counters value in CLI.
Added target oids.
Write lock the target instance when starting it.
Converted target counters to perf items.

* 2009/12/22 Guy Martin <gmsoft@tuxicoman.be>
Make helper_ipv4 use the helper_resize_payload API.
Fix typo Mhz -> mHz in input_docsis.
Don't try to stop the input when exiting if already stopped.
Don't enable snmpagent by default but only when -S is provided.
Added debug message when waiting for threads to finish.
Added dist-hook to save date of make distcheck in release.h.

* 2009/12/21 Guy Martin <gmsoft@tuxicoman.be>
Added DOCSIS 3 support.
Added match_docsis_mgmt.

* 2009/12/12 Guy Martin <gmsoft@tuxicoman.be>
Check for net-snmp headers as well at configure time.

conf.c
/*
 *  packet-o-matic : modular network traffic processor
 *  Copyright (C) 2006 Guy Martin <gmsoft@tuxicoman.be>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#include <libxml/xmlmemory.h>
#include <libxml/parser.h>

#include "conf.h"
#include "input.h"
#include "target.h"
#include "match.h"
#include "conntrack.h"
#include "helper.h"

struct conf *config_alloc() {

	struct conf *c;
	c = malloc(sizeof(struct conf));
	bzero(c, sizeof(struct conf));
	return c;

}

int config_cleanup(struct conf* c) {

	input_cleanup(c->input);
	list_destroy(c->rules);

	free(c);
	return 1;
}

struct input* config_parse_input(xmlDocPtr doc, xmlNodePtr cur) {
	char *input_type;
	input_type = (char*) xmlGetProp(cur, (const xmlChar*) "type");
	if (!input_type) {
		dprint("No type given in the input tag\n");
		return NULL;
	}
	dprint("Parsing input of type %s\n", input_type);
	int it = input_register(input_type);
	if (it == -1) {
		dprint("Input %s not supported\n", input_type);
		xmlFree(input_type);
		return NULL;
	}
	struct input *ip = input_alloc(it);
	xmlNodePtr pcur = cur->xmlChildrenNode;
	while (pcur) {
		if (!xmlStrcmp(pcur->name, (const xmlChar*) "param")) {
			char *param_type = (char *) xmlGetProp(pcur, (const xmlChar*) "name");
			if (!param_type)
				continue;
			char *value = (char *) xmlNodeListGetString(doc, pcur->xmlChildrenNode, 1);
			if (!value) {
				xmlFree(param_type);
				continue;
			}
			if (!input_set_param(ip, param_type, value))
				dprint("No parameter %s for input %s\n", param_type, input_type);

			xmlFree(param_type);
			xmlFree(value);

		}
		pcur = pcur->next;
	}
	xmlFree(input_type);

	return ip;
}

struct target *parse_target(xmlDocPtr doc, xmlNodePtr cur) {

	
	char *target_type;
	target_type = (char*) xmlGetProp(cur, (const xmlChar*) "type");
	if (!target_type) {
		dprint("No type given in the target tag\n");
		return NULL;
	}
	dprint("Parsing target of type %s\n", target_type);
	int tt = target_register(target_type);
	if (tt == -1) {
		dprint("Target %s not supported\n", target_type);
		xmlFree(target_type);
		return NULL;
	}
	struct target *tp = target_alloc(tt);

	if (!tp) {
		dprint("Error, unable to allocate target of type %s\n", target_type);
		xmlFree(target_type);
		return NULL;
	}

	xmlNodePtr pcur = cur->xmlChildrenNode;
	while (pcur) {
		if (!xmlStrcmp(pcur->name, (const xmlChar*) "param")) {
			char *param_type = (char *) xmlGetProp(pcur, (const xmlChar*) "name");
			if (!param_type)
				continue;
			char *value = (char *) xmlNodeListGetString(doc, pcur->xmlChildrenNode, 1);
			if (!value) {
				xmlFree(param_type);
				continue;
			}
			if (!target_set_param(tp, param_type, value))
				dprint("No parameter %s for target %s\n", param_type, target_type);

			xmlFree(param_type);
			xmlFree(value);

		}
		pcur = pcur->next;
	}
	xmlFree(target_type);

	target_open(tp);

	return tp;



}

struct rule_node *parse_match(xmlDocPtr doc, xmlNodePtr cur) {


	if (!cur)
		return NULL;




	while (cur) {
		if (!xmlStrcmp(cur->name, (const xmlChar *) "match")) {
			
			
			char *match_type = (char *) xmlGetProp(cur, (const xmlChar*) "type");
			if (!match_type) {
				dprint("No type given in the match tag\n");
				return NULL;
			}
			dprint("Parsing match of type %s\n", match_type);
			int mt = match_register(match_type);
			if (mt == -1) {
				dprint("Match %s not supported\n", match_type);
				xmlFree(match_type);
				return NULL;
			}
			struct match *mp = match_alloc(mt);

			struct rule_node *n;
			n = malloc(sizeof(struct rule_node));
			bzero(n, sizeof(struct rule_node));
			ndprint("Creating new rule_node\n");

			n->match = mp;

			xmlNodePtr pcur = cur->xmlChildrenNode;
			while (pcur) {
				if (!xmlStrcmp(pcur->name, (const xmlChar*) "param")) {
					char *param_type = (char *) xmlGetProp(pcur, (const xmlChar*) "name");
					if (!param_type)
						continue;
					char *value = (char *) xmlNodeListGetString(doc, pcur->xmlChildrenNode, 1);
					if (!value) {
						xmlFree(param_type);
						continue;
					}
					if (!match_set_param(mp, param_type, value))
						dprint("No parameter %s for match %s\n", param_type, match_type);

					xmlFree(param_type);
					xmlFree(value);

				}
				pcur = pcur->next;
			}

			// Try to register corresponding conntrack and helper
			conntrack_register(match_type);
			helper_register(match_type);

			xmlFree(match_type);

			n->a = parse_match(doc, cur->next);
			return n;
		} else if (!xmlStrcmp(cur->name, (const xmlChar *) "node")) {
			xmlChar* op = xmlGetProp(cur, (const xmlChar*) "op");
			if (!op) {
				dprint("No op specified in node tag\n");
				return NULL;
			}

			struct rule_node *n;
			n = malloc(sizeof(struct rule_node));
			bzero(n, sizeof(struct rule_node));
			ndprint("Creating new rule_node\n");

			if (!xmlStrcmp(op, (const xmlChar*) "or"))
				n->andor = RULE_OP_OR;
			else if (!xmlStrcmp(op, (const xmlChar*) "and" ))
				n->andor = RULE_OP_AND;
			else {
				dprint("Invalid operation %s for node\n", op);
				xmlFree(op);
				free(n);
				return NULL;
			}

			xmlFree(op);


			xmlNodePtr pcur = cur->xmlChildrenNode;

			while (!pcur) {
				if (!xmlStrcmp(cur->name, (const xmlChar *) "node") || !xmlStrcmp(cur->name, (const xmlChar *) "match")) {
					if (!n->a)
						n->a = parse_match(doc, pcur);
					else if (!n->b)
						n->b = parse_match(doc, pcur);
					else {
						dprint ("Too many sub nodes in node\n");
						return NULL;
					}
				}
				pcur = pcur->next;
						
			}
		
			return n;
		}
		cur = cur->next;

	}

	return NULL;

}

struct rule_list *parse_rule(xmlDocPtr doc, xmlNodePtr cur) {


	struct rule_list *r;
	r = malloc(sizeof(struct rule_list));
	bzero(r, sizeof(struct rule_list));
	
	cur = cur->xmlChildrenNode;

	while (cur) {
		if (!xmlStrcmp(cur->name, (const xmlChar *) "target")) {
			if (r->target)
				dprint("Only one target supported. Skipping extra target defined\n");
			else
				r->target = parse_target(doc, cur);
		} else if (!xmlStrcmp(cur->name, (const xmlChar *) "matches")) {
			if (r->node)
				dprint("Only one instance of matches supported. Skipping extra instances\n");
			else
				r->node = parse_match(doc, cur->xmlChildrenNode);
		}

		cur = cur->next;
	}

	return r;

}


int config_parse(struct conf *c, char * filename) {

	
	xmlDocPtr doc;
	xmlNodePtr root, cur;

	doc = xmlParseFile(filename);

	if (!doc) {
		dprint("Parse error when parsing %s!\n", filename);
		return 0;
	}

	root = xmlDocGetRootElement(doc);

	if (!root) {
		dprint("Hey dude, ya better gimme a non empty config file !\n");
		xmlFreeDoc(doc);
		return 0;
	}

	if (xmlStrcmp(root->name, (const xmlChar *) "config")) {
		dprint("The first node should be config !\n");
		return 0;
	}

	cur = root->xmlChildrenNode;

	while (cur) {
		if (!xmlStrcmp(cur->name, (const xmlChar *) "input")) {
			if (c->input)
				dprint("Only one input supported. Skipping extra input defined\n");
			else
				c->input = config_parse_input(doc, cur);
		} else if (!xmlStrcmp(cur->name, (const xmlChar *) "rule")) {
			struct rule_list *r = parse_rule(doc, cur);
			r->next = c->rules;
			c->rules = r;
		}

		cur = cur->next;
	}



	xmlFreeDoc(doc);

	xmlCleanupParser();

	return 1;
}
